### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by jranks123**
. Pulls out the creation of the identityID from the trackContribution function
. models PaypalAPI error as a PaypalBackendError

### CODE REVIEW

File: src/main/scala/backend/PaypalBackend.scala [hunk 1]
```diff
@@ -57,18 +62,11 @@ class PaypalBackend(
...
-      }
-      .flatMap { identityId =>
-        ContributionData.fromPaypalCharge(identityId, payment)
-          .leftMap(_ => PaypalBackendError.fromDatabaseError(null))
-          .toEitherT[Future]
-      }
+
+  def trackContribution(payment: Payment, acquisitionData: AcquisitionData, identityId: Option[Long])(implicit pool: DefaultThreadPool): EitherT[Future, PaypalBackendError, Unit]  = {
+    ContributionData.fromPaypalCharge(identityId, payment)
+      .leftMap(_ => PaypalBackendError.fromDatabaseError(null))
```
Comment (Mullefa) [inline src/main/scala/backend/PaypalBackend.scala]: this wasn't introduced in your pull request, but could you fix the `leftMap`?

File: src/main/scala/backend/PaypalBackend.scala [hunk 2]
```diff
@@ -86,28 +84,35 @@ class PaypalBackend(
...
    *  Use by Apps: Apps have previously created the payment and managed its approval with the customer.
    *  Funds are captured at this stage.
    */
   def capturePayment(capturePaypalPaymentData: CapturePaypalPaymentData)(implicit pool: DefaultThreadPool):
-  EitherT[Future, PaypalApiError, Payment] = {
+  EitherT[Future, PaypalBackendError, Payment] = {
     for {
-      payment <- paypalService.capturePayment(capturePaypalPaymentData)
-      _ = trackContribution(payment, capturePaypalPaymentData.acquisitionData, None)
+      payment <- paypalService.capturePayment(capturePaypalPaymentData).leftMap(PaypalBackendError.fromPaypalAPIError)
```
Comment (Mullefa) [inline src/main/scala/backend/PaypalBackend.scala]: I _think_ it's probably better to have these left maps outside of the `for` comprehension - it reduces noise and emphasises this a monadic flow which could work with any monad, at the cost of having to create a utility method.

File: src/main/scala/backend/PaypalBackend.scala [hunk 3]
```diff
@@ -86,28 +84,35 @@ class PaypalBackend(
...
+  def processPayment(paypalExecutePaymentData: ExecutePaypalPaymentData)(implicit pool: DefaultThreadPool):
+  EitherT[Future, PaypalBackendError, Payment] = {
     for {
-      payment <- paypalService.executePayment(paypalExecutePaymentData)
-      _ = trackContribution(payment, paypalExecutePaymentData.acquisitionData, paypalExecutePaymentData.signedInUserEmail)
+      payment <- executePayment(paypalExecutePaymentData)
+
+      signedInUserEmail = paypalExecutePaymentData.signedInUserEmail
+      paymentEmail = payment.getPayer.getPayerInfo.getEmail
+      email = signedInUserEmail.getOrElse(paymentEmail)
```
Comment (Mullefa) [inline src/main/scala/backend/PaypalBackend.scala]: as discussed, it's worth thinking through all the implications of recording the contribution against an email different to the one associated with the Paypal account used to make it

**Comment by jbguardian** [comment]
LGTM 👍  Although I'm not a big fan of having - PaypalBackendError - StripeChargeError - PaypalApiError. 
I see the point of having the payPalBackendSemiGroup and I think it's a great idea but I don't see the point of having 3 different error classes with almost same structure which will end up in the same error code in the controller
We could replace them by only one Error class making easier to mix them in the Backend classes and removing .fromOphanError, .fromDatabaseError and so on.
Probably out of the scope of this PR.

**Comment by Mullefa** [comment]
@jbguardian definitely merits a chat!

**Comment by joelochlann** [comment]
Nice suggestion @jbguardian, perhaps "Solution 2: ADTs all the way down" on https://typelevel.org/cats/datatypes/either.html is relevant to that discussion

**Pull request merged**